File: panel/src/main/java/com/effective/android/panel/PanelSwitchHelper.java
package com.effective.android.panel;

import android.app.Activity;
import android.support.annotation.IdRes;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.Fragment;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;

import com.effective.android.panel.interfaces.OnScrollOutsideBorder;
import com.effective.android.panel.interfaces.listener.OnEditFocusChangeListener;
import com.effective.android.panel.interfaces.listener.OnKeyboardStateListener;
import com.effective.android.panel.interfaces.listener.OnPanelChangeListener;
import com.effective.android.panel.interfaces.listener.OnViewClickListener;
import com.effective.android.panel.utils.PanelUtil;
import com.effective.android.panel.view.PanelSwitchLayout;

import java.util.ArrayList;
import java.util.List;

/**
 * the helper of panel switching
 * Created by yummyLau on 2018-6-21.
 * Email: yummyl.lau@gmail.com
 * blog: yummylau.com
 * <p>
 * updated by yummyLau on 20/03/18
 * 重构整个输入法切换框架，移除旧版使用 weight+Runnable延迟切换，使用新版 layout+动画无缝衔接！
 */

public final class PanelSwitchHelper {

    private PanelSwitchLayout mPanelSwitchLayout;
    private boolean canScrollOutside;

    private PanelSwitchHelper(Builder builder) {
        Constants.DEBUG = builder.logTrack;
        this.canScrollOutside = builder.contentCanScrollOutside;
        if (builder.logTrack) {
            builder.viewClickListeners.add(LogTracker.getInstance());
            builder.panelChangeListeners.add(LogTracker.getInstance());
            builder.keyboardStatusListeners.add(LogTracker.getInstance());
            builder.editFocusChangeListeners.add(LogTracker.getInstance());
        }
        mPanelSwitchLayout = builder.panelSwitchLayout;
        mPanelSwitchLayout.setScrollOutsideBorder(new OnScrollOutsideBorder() {
            @Override
            public boolean canLayoutOutsideBorder() {
                return canScrollOutside;
            }

            @Override
            public int getOutsideHeight() {
                return PanelUtil.getKeyBoardHeight(mPanelSwitchLayout.getContext());
            }
        });
        mPanelSwitchLayout.bindListener(builder.viewClickListeners, builder.panelChangeListeners, builder.keyboardStatusListeners, builder.editFocusChangeListeners);
        mPanelSwitchLayout.bindWindow(builder.window);
    }

    public boolean hookSystemBackByPanelSwitcher() {
        return mPanelSwitchLayout.hookSystemBackByPanelSwitcher();
    }

    public void scrollOutsideEnable(boolean enable){
        this.canScrollOutside = enable;
    }


    /**
     * 外部显示输入框
     */
    public void toKeyboardState() {
        mPanelSwitchLayout.toKeyboardState();
    }

    /**
     * 隐藏输入法或者面板
     */
    public void resetState() {
        mPanelSwitchLayout.checkoutPanel(Constants.PANEL_NONE);
    }


    public static class Builder {

        List<OnViewClickListener> viewClickListeners;
        List<OnPanelChangeListener> panelChangeListeners;
        List<OnKeyboardStateListener> keyboardStatusListeners;
        List<OnEditFocusChangeListener> editFocusChangeListeners;
        PanelSwitchLayout panelSwitchLayout;
        Window window;
        View rootView;
        boolean logTrack;
        boolean contentCanScrollOutside = true;

        public Builder(@NonNull Activity activity) {
            this(activity.getWindow(), activity.getWindow().getDecorView().findViewById(android.R.id.content));
        }

        public Builder(@NonNull Fragment fragment) {
            this(fragment.getActivity().getWindow(), fragment.getView());
        }

        public Builder(@NonNull DialogFragment dialogFragment) {
            this(dialogFragment.getActivity().getWindow(), dialogFragment.getView());
        }

        public Builder(@NonNull Window window, @NonNull View root) {
            this.window = window;
            this.rootView = root;
            viewClickListeners = new ArrayList<>();
            panelChangeListeners = new ArrayList<>();
            keyboardStatusListeners = new ArrayList<>();
            editFocusChangeListeners = new ArrayList<>();
        }

        /**
         * note: helper will set view's onClickListener to View ,so you should add OnViewClickListener for your project.
         *
         * @param listener
         * @return
         */
        public Builder addViewClickListener(OnViewClickListener listener) {
            if (listener != null && !viewClickListeners.contains(listener)) {
                viewClickListeners.add(listener);
            }
            return this;
        }

        public Builder addPanelChangeListener(OnPanelChangeListener listener) {
            if (listener != null && !panelChangeListeners.contains(listener)) {
                panelChangeListeners.add(listener);
            }
            return this;
        }

        public Builder addKeyboardStateListener(OnKeyboardStateListener listener) {
            if (listener != null && !keyboardStatusListeners.contains(listener)) {
                keyboardStatusListeners.add(listener);
            }
            return this;
        }

        public Builder addEditTextFocusChangeListener(OnEditFocusChangeListener listener) {
            if (listener != null && !editFocusChangeListeners.contains(listener)) {
                editFocusChangeListeners.add(listener);
            }
            return this;
        }

        public Builder contentCanScrollOutside(boolean canScrollOutside){
            this.contentCanScrollOutside = canScrollOutside;
            return this;
        }

        public Builder logTrack(boolean logTrack) {
            this.logTrack = logTrack;
            return this;
        }

        public PanelSwitchHelper build(boolean showKeyboard) {

            if (window == null) {
                throw new IllegalArgumentException("PanelSwitchHelper$Builder#build : window can't be null!please set value by call #Builder");
            }

            if (rootView == null) {
                throw new IllegalArgumentException("PanelSwitchHelper$Builder#build : rootView can't be null!please set value by call #Builder");
            }

            findSwitchLayout(rootView);
            if (panelSwitchLayout == null) {
                throw new IllegalArgumentException("PanelSwitchHelper$Builder#build : not found PanelSwitchLayout!");
            }

            final PanelSwitchHelper panelSwitchHelper = new PanelSwitchHelper(this);
            if (showKeyboard) {
                panelSwitchLayout.toKeyboardState();
            }
            return panelSwitchHelper;
        }

        public PanelSwitchHelper build() {
            return build(false);
        }

        private void findSwitchLayout(View view) {
            if (view instanceof PanelSwitchLayout) {
                if(panelSwitchLayout != null){
                    throw new IllegalArgumentException("PanelSwitchHelper$Builder#build : rootView has one more panelSwitchLayout!");
                }
                panelSwitchLayout = (PanelSwitchLayout) view;
                return;
            }

            if(view instanceof ViewGroup){
                int childCount = ((ViewGroup) view).getChildCount();
                for(int i = 0; i <  childCount; i++){
                    findSwitchLayout(((ViewGroup) view).getChildAt(i));
                }
            }
        }
    }
}

File: panel-kotlin/src/main/java/com/effective/android/panel/PanelSwitchHelper.kt
package com.effective.android.panel

import android.R
import android.app.Activity
import android.support.v4.app.DialogFragment
import android.support.v4.app.Fragment
import android.view.View
import android.view.ViewGroup
import android.view.Window
import com.effective.android.panel.interfaces.OnScrollOutsideBorder
import com.effective.android.panel.interfaces.listener.*
import com.effective.android.panel.utils.PanelUtil.getKeyBoardHeight
import com.effective.android.panel.view.PanelSwitchLayout

/**
 * the helper of panel switching
 * Created by yummyLau on 2018-6-21.
 * Email: yummyl.lau@gmail.com
 * blog: yummylau.com
 *
 *
 * updated by yummyLau on 20/03/18
 * 重构整个输入法切换框架，移除旧版使用 weight+Runnable延迟切换，使用新版 layout+动画无缝衔接！
 */
class PanelSwitchHelper private constructor(builder: Builder,showKeyboard: Boolean) {

    private val mPanelSwitchLayout: PanelSwitchLayout
    private var canScrollOutside: Boolean

    init {
        Constants.DEBUG = builder.logTrack
        canScrollOutside = builder.contentCanScrollOutside
        if (builder.logTrack) {
            builder.viewClickListeners.add(LogTracker)
            builder.panelChangeListeners.add(LogTracker)
            builder.keyboardStatusListeners.add(LogTracker)
            builder.editFocusChangeListeners.add(LogTracker)
        }
        mPanelSwitchLayout = builder.panelSwitchLayout!!
        mPanelSwitchLayout.setScrollOutsideBorder(object : OnScrollOutsideBorder {
            override fun canLayoutOutsideBorder(): Boolean {
                return canScrollOutside
            }

            override fun getOutsideHeight(): Int = getKeyBoardHeight(mPanelSwitchLayout.context)
        })
        mPanelSwitchLayout.bindListener(builder.viewClickListeners, builder.panelChangeListeners, builder.keyboardStatusListeners, builder.editFocusChangeListeners)
        mPanelSwitchLayout.bindWindow(builder.window)
        if(showKeyboard){
            mPanelSwitchLayout.toKeyboardState()
        }
    }

    fun hookSystemBackByPanelSwitcher(): Boolean {
        return mPanelSwitchLayout.hookSystemBackByPanelSwitcher()
    }

    fun scrollOutsideEnable(enable: Boolean) {
        canScrollOutside = enable
    }

    /**
     * 外部显示输入框
     */
    fun toKeyboardState() {
        mPanelSwitchLayout.toKeyboardState()
    }

    /**
     * 隐藏输入法或者面板
     */
    fun resetState() {
        mPanelSwitchLayout.checkoutPanel(Constants.PANEL_NONE)
    }

    class Builder(window: Window?, root: View?) {
        var viewClickListeners: MutableList<OnViewClickListener> = mutableListOf()
        var panelChangeListeners: MutableList<OnPanelChangeListener> = mutableListOf()
        var keyboardStatusListeners: MutableList<OnKeyboardStateListener> = mutableListOf()
        var editFocusChangeListeners: MutableList<OnEditFocusChangeListener> = mutableListOf()
        var panelSwitchLayout: PanelSwitchLayout? = null
        var window: Window
        var rootView: View
        var logTrack = false
        var contentCanScrollOutside = true

        constructor(activity: Activity) : this(activity.window, activity.window.decorView.findViewById<View>(R.id.content))
        constructor(fragment: Fragment) : this(fragment.activity?.window, fragment.view)
        constructor(dialogFragment: DialogFragment) : this(dialogFragment.activity?.window, dialogFragment.view)

        init {
            requireNotNull(window) { "PanelSwitchHelper\$Builder#build : window can't be null!please set value by call #Builder" }
            this.window = window
            requireNotNull(root) { "PanelSwitchHelper\$Builder#build : rootView can't be null!please set value by call #Builder" }
            this.rootView = root
        }

        /**
         * note: helper will set view's onClickListener to View ,so you should add OnViewClickListener for your project.
         *
         * @param listener
         * @return
         */
        fun addViewClickListener(listener: OnViewClickListener): Builder {
            if (!viewClickListeners.contains(listener)) {
                viewClickListeners.add(listener)
            }
            return this
        }

        fun addViewClickListener(function: OnViewClickListenerBuilder.() -> Unit): Builder {
            viewClickListeners.add(OnViewClickListenerBuilder().also(function))
            return this
        }

        fun addPanelChangeListener(listener: OnPanelChangeListener): Builder {
            if (!panelChangeListeners.contains(listener)) {
                panelChangeListeners.add(listener)
            }
            return this
        }

        fun addPanelChangeListener(function: OnPanelChangeListenerBuilder.() -> Unit): Builder {
            panelChangeListeners.add(OnPanelChangeListenerBuilder().also(function))
            return this
        }

        fun addKeyboardStateListener(listener: OnKeyboardStateListener): Builder {
            if (!keyboardStatusListeners.contains(listener)) {
                keyboardStatusListeners.add(listener)
            }
            return this
        }

        fun addKeyboardStateListener(function: OnKeyboardStateListenerBuilder.() -> Unit): Builder {
            keyboardStatusListeners.add(OnKeyboardStateListenerBuilder().also(function))
            return this
        }

        fun addEditTextFocusChangeListener(listener: OnEditFocusChangeListener): Builder {
            if (!editFocusChangeListeners.contains(listener)) {
                editFocusChangeListeners.add(listener)
            }
            return this
        }

        fun addEditTextFocusChangeListener(function: OnEditFocusChangeListenerBuilder.() -> Unit): Builder {
            editFocusChangeListeners.add(OnEditFocusChangeListenerBuilder().also(function))
            return this
        }

        fun contentCanScrollOutside(canScrollOutside: Boolean): Builder {
            contentCanScrollOutside = canScrollOutside
            return this
        }

        fun logTrack(logTrack: Boolean): Builder {
            this.logTrack = logTrack
            return this
        }

        @JvmOverloads
        fun build(showKeyboard: Boolean = false): PanelSwitchHelper {
            findSwitchLayout(rootView)
            requireNotNull(panelSwitchLayout) { "PanelSwitchHelper\$Builder#build : not found PanelSwitchLayout!" }
            return PanelSwitchHelper(this, showKeyboard)
        }

        private fun findSwitchLayout(view: View) {
            if (view is PanelSwitchLayout) {
                require(panelSwitchLayout == null) { "PanelSwitchHelper\$Builder#build : rootView has one more panelSwitchLayout!" }
                panelSwitchLayout = view
                return
            }
            if (view is ViewGroup) {
                val childCount = view.childCount
                for (i in 0 until childCount) {
                    findSwitchLayout(view.getChildAt(i))
                }
            }
        }
    }
}